### PR TITLE
[R] Move dialogs when keyboard pops up

### LINKED ISSUE #7032: Dialogs don't get out of the way of system padding (e.g. from on-screen keyboards)
I pop a dialog with a single autofocussing text field.
- ~~The keyboard doesn't pop up (it does in e.g. the android system settings wifi "add network" dialog)~~ https://github.com/flutter/flutter/issues/7035
- ~~Once I pop the keyboard up, the text behind the dialog all disappears. I'm using nested Scaffolds, which is likely the issue.~~ https://github.com/flutter/flutter/issues/7036
- The dialog doesn't move (and so gets covered by the keyboard).
- ~~Also the text field uses my primary color and my buttons use the accent color. The dialog has no black at all.~~ That's what the spec says. I added a title and it looks ok.
- ~~Even the cursor is the primary color, unlike the text itself, which is black. this looks weird.~~ That's what the spec says.

code to show the dialog:
```dart
  Future<String> showMessageDialog() {
    String result;
    return showDialog(
      context: context,
      child: new Form(
        child: new Builder(
          builder: (BuildContext context) {
            return new AlertDialog(
              content: new Block(
                children: <Widget>[
                  new InputFormField(
                    labelText: 'Message',
                    autofocus: true,
                    onSaved: (InputValue value) {
                      result = value.text;
                    },
                  ),
                ],
              ),
              actions: <Widget>[
                new FlatButton(
                  onPressed: () {
                    Navigator.pop(context);
                  },
                  child: new Text('CANCEL'),
                ),
                new FlatButton(
                  onPressed: () {
                    Form.of(context).save();
                    Navigator.pop(context, result);
                  },
                  child: new Text('SEND'),
                ),
              ],
            );
          }
        ),
      ),
    );
  }
```

The app has a Scaffold containing a Block. The dialog above is triggered from a FlatButton in that Block.

### PR DESCRIPTION
**Pull request by Hixie**
Fixes #7032 

cc @cbracken

### CODE REVIEW

File: packages/flutter/lib/src/material/dialog.dart [hunk 1]
```diff
@@ -38,32 +39,55 @@ class Dialog extends StatelessWidget {
...
 
   /// The widget below this widget in the tree.
   ///
   /// {@macro flutter.widgets.child}
   final Widget child;
 
+  /// The duration of the animation to show when the system keyboard intrudes
+  /// into the space that the dialog is placed in.
+  ///
+  /// Defaults to 200 milliseconds.
```
Comment (cbracken) [inline packages/flutter/lib/src/material/dialog.dart]: Should it be 100 or 200 ms? The constructor code default is currently 100.

Comment (cbracken) [reply packages/flutter/lib/src/material/dialog.dart]: Which according to our in-person conversation right now is the correct value :)

File: packages/flutter/test/material/dialog_test.dart [hunk 2]
```diff
@@ -262,15 +264,62 @@ void main() {
...
+          viewInsets: const EdgeInsets.fromLTRB(10.0, 20.0, 30.0, 40.0),
+        ),
+        child: const Dialog(
+          child: const Placeholder(),
+        ),
+      ),
+    );
+    expect(
+      tester.getRect(find.byType(Placeholder)),
+      new Rect.fromLTRB(10.0 + 40.0, 20 + 24.0, 800.0 - (40.0 + 30.0), 600.0 - (24.0 + 40.0)),
```
Comment (cbracken) [inline packages/flutter/test/material/dialog_test.dart]: `s/20/20.0/` in the second argument.

**Review by cbracken — APPROVED** [review]

**Comment by cbracken** [comment]
LGTM modulo comments.
lgtm

**Comment by Hixie** [comment]
Thanks. Will land on green.

**Comment by FrankenApps** [comment]
Well, you fixed it for the material dialog, but should the cupertino dialog not behave the same?

**Pull request merged**